src/core/electrostatics_magnetostatics/icc.cpp

```diff
@@ -72,11 +72,14 @@ void force_calc_iccp3m(const ParticleRange &particles,
 inline void add_non_bonded_pair_force_iccp3m(Particle *p1, Particle *p2,
                                              Utils::Vector3d const &d,
                                              double dist, double dist2) {
-  Utils::Vector3d force{};
-  Coulomb::calc_pair_force(p1, p2, d, dist, force);
+  auto forces = Coulomb::calc_pair_force(p1, p2, d, dist);
 
-  p1->f.f += force;
-  p2->f.f -= force;
+  p1->f.f += std::get<0>(forces);
+  p2->f.f -= std::get<0>(forces);
+#ifdef P3M
+  p1->f.f += std::get<1>(forces);
+  p2->f.f += std::get<2>(forces);
+#endif
```

Comment:
This still modifies the particle. And I think we also add this force down in forces_inline.hpp.

Reply:
True, it's not a pure function, but it's also not a force kernel. It's only a helper function that is not visible outside this file. The forces here are obtained from `Coulomb::pair_force(p1, p2, d, dist);`, which is indeed called in `forces_inline.hpp`. I might be wrong, but I think this helper function is used by [`iccp3m_iteration()`](https://github.com/espressomd/espresso/blob/54bbb885120e166c0a6945ef196bcacabf10a971/src/core/electrostatics_magnetostatics/icc.cpp#L94) to progressively update virtual charges while making sure the resulting particle force doesn't cancel the external electric field. Unfortunately this function isn't documented.